### ISSUE TITLE
[BUG] `cython --working` isn't used for relative cimports

### ISSUE THREAD

**Issue by godlygeek** [cython/cython#3862]
**Describe the bug**
When using the `cython` CLI to compile a `.pyx` file in another directory, `.pxd` and `__init__.py` files for relative a `cimport` are looked up in the directory containing the `.pyx`, rather than the `--working` directory.

**To Reproduce**
Code to reproduce the behaviour:

In `test.pyx`:
```cython
from . cimport other

def test():
    other.puts("hello, world!")
```

In `other.pxd`:
```cython
cdef extern from "stdio.h":
    int puts(const char *)
```

And create an empty `__init__.py`:
```console
touch __init__.py
```

From this directory, `cython test.pyx` works as expected and produces a `test.c`.

But, if you `mv test.pyx /tmp` and then instead run:
```
cython --working $PWD /tmp/test.pyx
```
you get an error:
```
from . cimport other
^
------------------------------------------------------------

/tmp/test.pyx:1:0: relative cimport beyond main package is not allowed
```

**Expected behavior**
I'm not exactly sure whether this is a bug or not.  I'm using an editor plugin that wants to automatically compile Cython code to lint it and find errors.  It does this by copying the not-yet-saved file from the editor into a temporary directory, and then using `--working` to point it at the rest of the code.  This fails for anything that's using relative imports, though, because Cython searches for the relative cimports relative to the `.pyx` file it's compiling, rather than relative to the `--working` directory that it was given, and I don't see any switches to change that behavior.

Should this work?  If not, is there another way to compile a `.pyx` file and pretend it came from a different directory?

**Environment (please complete the following information):**
 - OS: Linux
 - Python version: 3.8.6
 - Cython version: 0.29.21

**Comment by scoder** [comment]
Thanks for the report. I do not think what you propose is how it should work. The working directory is where Cython puts its generated code, not where it looks for source files.

I see two related things, but they probably won't help you either. One is the use of [namespace packages](https://packaging.python.org/guides/packaging-namespace-packages/), which Cython (3?) supports now. The other is making sure that the current work directory is at least on the Python package search path, so that `.pxd` files can be looked up there. But that won't help with relative imports.

Basically, the question is whether the expectation of that plugin is correct for `.py` files in Python, and I think the answer is No. Then it's correct for the answer in Cython to also be No.

**Comment by godlygeek** [comment]
Perhaps an alternative would be for `cython` to support taking the code to transpile from stdin rather than a file on disk, and locating other files relative to it based on the working directory? That essentially is how Python would locate modules if you ran `import foo | python`, though admittedly that's not a perfect comparison because that wouldn't support relative imports - for those you'd need to use `python -m`.

But perhaps reading from stdin would let the plugin avoid the temp file, and directory, entirely.

**Comment by scoder** [comment]
Why couldn't the plugin create a temp file in the source directory? Like many editors do?

**Comment by godlygeek** [comment]
Since I'm not the author I'd be guessing at the motivations, but my best guess is that by using `$TMPDIR` it's likely to get cleaned up eventually even if the process crashes, and not risk leaving something around in the user's working copy that tools might pick up and get confused by.

Similar linters for gcc, for example, do pipe the code on stdin to avoid the need for a temp file, which seems reasonable to me as it should be both faster and more reliable.

**Comment by godlygeek** [comment]
> The working directory is where Cython puts its generated code, not where it looks for source files.

This seems pretty definitive in favor of saying this isn't a bug, and we're in feature request territory.

I suppose just reading from stdin isn't enough for the feature I need... You need the ability not just to get the contents of the .pyx file, but also to locate other files relative to it, and using `$PWD` for that may also be unexpected. Maybe the feature I'm looking for is to compile a given file as though it's in some given directory, like I assumed `--working` did.
